File: src/solver/subroutines/solveFinalEdges.ts
import { cubesShareFace, positionMap } from '../../utils'
import RubiksCubeSolver from '../RubiksCubeSolver'

export default async function solveFinalEdges(solver: RubiksCubeSolver) {
  const leftEdge = solver.cubeMap.get(positionMap[4])
  const rightEdge = solver.cubeMap.get(positionMap[6])
  const frontEdge = solver.cubeMap.get(positionMap[8])
  const frontRight = solver.cubeMap.get(positionMap[9])
  const backLeft = solver.cubeMap.get(positionMap[1])
  const backRight = solver.cubeMap.get(positionMap[3])
  const frontLeft = solver.cubeMap.get(positionMap[7])

  if (cubesShareFace('left', frontLeft, leftEdge, backLeft)) {
    await solver.do('XCW')
  } else if (cubesShareFace('right', frontRight, rightEdge, backRight)) {
    await solver.do('XCCW')
  } else if (cubesShareFace('front', frontLeft, frontEdge, frontRight)) {
    await solver.do('XCW', 'XCW')
  }

  while (
    solver.cubeMap.get(positionMap[1])?.orientation.back !==
    solver.cubeMap.get(positionMap[11])?.orientation.back
  ) {
    await solver.do('rotateTopCW', 'XCCW')
  }

  if (leftEdge?.orientation.left === solver.cubeMap.get(positionMap[17])?.orientation.front) {
    await runAlgorithmLR(solver)
    return
  } else {
    await runAlgorithmRL(solver)
    return
  }
}

async function runAlgorithmLR(solver: RubiksCubeSolver) {
  await solver.do(
    'rotateFrontCW',
    'rotateFrontCW',
    'rotateTopCCW',
    'rotateLeftCCW',
    'rotateRightCCW',
    'rotateFrontCW',
    'rotateFrontCW',
    'rotateLeftCW',
    'rotateRightCW',
    'rotateTopCCW',
    'rotateFrontCW',
    'rotateFrontCW'
  )
}

async function runAlgorithmRL(solver: RubiksCubeSolver) {
  await solver.do(
    'rotateFrontCW',
    'rotateFrontCW',
    'rotateTopCW',
    'rotateLeftCCW',
    'rotateRightCCW',
    'rotateFrontCW',
    'rotateFrontCW',
    'rotateLeftCW',
    'rotateRightCW',
    'rotateTopCW',
    'rotateFrontCW',
    'rotateFrontCW'
  )
}

File: android/core/src/main/kotlin/com/mikeb/simplepuzzlecube/solver/subroutines/solveFinalEdges.kt
// Port of ../src/solver/subroutines/solveFinalEdges.ts.
package com.mikeb.simplepuzzlecube.solver.subroutines

import com.mikeb.simplepuzzlecube.cubesShareFace
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateFrontCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateLeftCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateLeftCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateRightCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateRightCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateTopCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateTopCW
import com.mikeb.simplepuzzlecube.engine.OrientationKey
import com.mikeb.simplepuzzlecube.engine.Rotation.XCCW
import com.mikeb.simplepuzzlecube.engine.Rotation.XCW
import com.mikeb.simplepuzzlecube.solver.RubiksCubeSolver

suspend fun solveFinalEdges(solver: RubiksCubeSolver) {
    val leftEdge = solver.fetchPosition(4)
    val rightEdge = solver.fetchPosition(6)
    val frontEdge = solver.fetchPosition(8)
    val frontRight = solver.fetchPosition(9)
    val backLeft = solver.fetchPosition(1)
    val backRight = solver.fetchPosition(3)
    val frontLeft = solver.fetchPosition(7)

    if (cubesShareFace(OrientationKey.left, frontLeft, leftEdge, backLeft)) {
        solver.`do`(XCW)
    } else if (cubesShareFace(OrientationKey.right, frontRight, rightEdge, backRight)) {
        solver.`do`(XCCW)
    } else if (cubesShareFace(OrientationKey.front, frontLeft, frontEdge, frontRight)) {
        solver.`do`(XCW, XCW)
    }

    while (solver.fetchPosition(1)?.orientation?.back != solver.fetchPosition(11)?.orientation?.back) {
        solver.`do`(rotateTopCW, XCCW)
    }

    if (leftEdge?.orientation?.left == solver.fetchPosition(17)?.orientation?.front) {
        runAlgorithmLR(solver)
        return
    } else {
        runAlgorithmRL(solver)
        return
    }
}

private suspend fun runAlgorithmLR(solver: RubiksCubeSolver) {
    solver.`do`(
        rotateFrontCW,
        rotateFrontCW,
        rotateTopCCW,
        rotateLeftCCW,
        rotateRightCCW,
        rotateFrontCW,
        rotateFrontCW,
        rotateLeftCW,
        rotateRightCW,
        rotateTopCCW,
        rotateFrontCW,
        rotateFrontCW
    )
}

private suspend fun runAlgorithmRL(solver: RubiksCubeSolver) {
    solver.`do`(
        rotateFrontCW,
        rotateFrontCW,
        rotateTopCW,
        rotateLeftCCW,
        rotateRightCCW,
        rotateFrontCW,
        rotateFrontCW,
        rotateLeftCW,
        rotateRightCW,
        rotateTopCW,
        rotateFrontCW,
        rotateFrontCW
    )
}
